runtime/bindings/python/iree/runtime/system_setup.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Process-global driver instantiation and discovery."""
+
+from typing import Collection, Dict, Optional, Sequence, Union
+
+import logging
+import os
+from threading import RLock
+
+from ._binding import _create_hal_driver, HalDevice, HalDriver
+
+_GLOBAL_DRIVERS = {}  # type: Dict[str, Union[HalDriver, Exception]]
+_GLOBAL_DEVICES_BY_NAME = {}  # type: Dict[str, Union[HalDevice, Exception]]
+
+_LOCK = RLock()
+
+DEFAULT_DRIVER_NAMES = "dylib,cuda,vulkan,vmvx"
+
+
+def query_available_drivers() -> Collection[str]:
+  """Returns a collection of driver names that are available."""
+  return HalDriver.query()
+
+
+def get_driver(driver_name: str) -> HalDriver:
+  """Gets or creates a driver by name."""
+  with _LOCK:
+    existing = _GLOBAL_DRIVERS.get(driver_name)
+    if existing is not None:
+      if isinstance(existing, Exception):
+        raise existing
+      else:
+        return existing
+
+    available_names = query_available_drivers()
+    if driver_name not in available_names:
+      raise ValueError(
+          f"Driver '{driver_name}' is not compiled into this binary")
+
+    try:
+      driver = _create_hal_driver(driver_name)
+    except Exception as ex:
+      _GLOBAL_DRIVERS[driver_name] = ex
+      raise
+    _GLOBAL_DRIVERS[driver_name] = driver
+    return driver
+
+
+def get_device_by_name(name_spec: str) -> HalDevice:
+  """Gets a cached device by name.
+
+  Args:
+    name_spec: The name of a driver or "{driver_name}:{index}" to create
+      a specific device. If the indexed form is not used, the driver
+      will be asked to create its default device.
```

Comment:
filed #9343 to sketch this out - I think it's compatible with the python API here but generalizes things to strings instead of IDs and that lets us make stable identifiers (UUIDs/LUIDs/etc).